rangeproof/rangeproof_test.go

```diff
@@ -0,0 +1,545 @@
+package rangeproof
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/privacybydesign/gabi/big"
+	"github.com/privacybydesign/gabi/internal/common"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGroup(t *testing.T) {
+	g := (qrGroup)(NewQrGroup(big.NewInt(35), big.NewInt(3), big.NewInt(4)))
+
+	assert.Equal(t, g.Base("R"), big.NewInt(3))
+	assert.Equal(t, g.Base("S"), big.NewInt(4))
+	assert.Equal(t, g.Base("N"), (*big.Int)(nil))
+	assert.Equal(t, g.Base("R1234"), (*big.Int)(nil))
+	assert.ElementsMatch(t, g.Names(), []string{"R", "S"})
+
+	ret := new(big.Int)
+	assert.True(t, g.Exp(ret, "R", big.NewInt(5), g.N))
+	assert.Equal(t, ret, new(big.Int).Exp(g.Base("R"), big.NewInt(5), g.N))
+	assert.True(t, g.Exp(ret, "S", big.NewInt(7), g.N))
+	assert.Equal(t, ret, new(big.Int).Exp(g.Base("S"), big.NewInt(7), g.N))
+	assert.False(t, g.Exp(ret, "N", big.NewInt(9), g.N))
+	assert.False(t, g.Exp(ret, "R1234", big.NewInt(11), g.N))
+}
+
+type bruteForce3 struct{}
+
+func (_ *bruteForce3) Split(delta *big.Int) ([]*big.Int, error) {
+	if !delta.IsInt64() {
+		panic("too big")
+	}
+
+	d := delta.Int64()
+
+	if d > 1e9 || d < 0 {
+		panic("too big")
+	}
+
+	for i := int64(0); i*i <= d; i++ {
+		for j := int64(0); i*i+j*j <= d; j++ {
+			for k := int64(0); i*i+j*j+k*k <= d; k++ {
+				if i*i+j*j+k*k == d {
+					return []*big.Int{big.NewInt(i), big.NewInt(j), big.NewInt(k)}, nil
+				}
+			}
+		}
+	}
+
+	panic("Not found")
+}
+
+func (_ *bruteForce3) SquareCount() int {
+	return 3
+}
+
+func (_ *bruteForce3) Ld() uint {
+	return 8
+}
+
+type bruteForce4 struct{}
+
+func (_ *bruteForce4) Split(delta *big.Int) ([]*big.Int, error) {
+	if !delta.IsInt64() {
+		panic("too big")
+	}
+
+	d := delta.Int64()
+
+	if d > 1e9 || d < 0 {
+		panic("too big")
+	}
+
+	for i := int64(0); i*i <= d; i++ {
+		for j := int64(0); i*i+j*j <= d; j++ {
+			for k := int64(0); i*i+j*j+k*k <= d; k++ {
+				for l := int64(0); i*i+j*j+k*k+l*l <= d; l++ {
+					if i*i+j*j+k*k+l*l == d {
+						return []*big.Int{big.NewInt(i), big.NewInt(j), big.NewInt(k), big.NewInt(l)}, nil
+					}
+				}
+			}
+		}
+	}
+
+	panic("Not found")
+}
+
+func (_ *bruteForce4) SquareCount() int {
+	return 4
+}
+
+func (_ *bruteForce4) Ld() uint {
+	return 8
+}
+
+func TestRangeProofBasic(t *testing.T) {
```

Comment:
This function looks completely identical to a number of other tests below, apart from different splitters. Can you reduce code duplication by introducing a helper, like in the splitter test file?

Reply:
Yes but would prefer not to as for tests it is generally considered better to keep context instead of producing dry code, and using helpers here would in my opinion significantly cloud clarity of what is exactly under test and what the test is doing.

Reply:
To be clear, I only mean the functions that are identical except for their splitter.

To me, like in the other test file, refactoring these to use a helper would increase clarity, not cloud it. As it is now, I had to read all of these functions carefully to determine that they are in fact identical, except for their splitter. The existence of a helper function taking a splitter parameter would immediately indicate that one thing is tested using multiple splitters.

Reply:
ok, refactored.